gflownet/envs/base.py

```diff
@@ -63,46 +67,227 @@ def __init__(
         self.proxy_state_format = proxy_state_format
-        self._true_density = None
-        self._z = None
-        self.action_space = []
-        self.eos = len(self.action_space)
+        # Log SoftMax function
         self.logsoftmax = torch.nn.LogSoftmax(dim=1)
-        # Assertions
-        assert self.reward_norm > 0
-        assert self.reward_beta > 0
-        assert self.min_reward > 0
-
-    def copy(self):
-        # return an instance of the environment
-        return self.__class__(**self.__dict__)
-
-    def set_energies_stats(self, energies_stats):
-        self.energies_stats = energies_stats
-
-    def set_reward_norm(self, reward_norm):
-        self.reward_norm = reward_norm
+        # Action space
+        self.action_space = self.get_action_space()
+        self.action_space_torch = torch.tensor(
+            self.action_space, device=self.device, dtype=self.float
+        )
+        self.action_space_dim = len(self.action_space)
+        # Max trajectory length
+        self.max_traj_length = self.get_max_traj_length()
+        # Policy outputs
+        self.fixed_policy_output = self.get_policy_output(fixed_distribution)
+        self.random_policy_output = self.get_policy_output(random_distribution)
+        self.policy_output_dim = len(self.fixed_policy_output)
+        self.policy_input_dim = len(self.state2policy())
 
     @abstractmethod
-    def get_actions_space(self):
+    def get_action_space(self):
         """
         Constructs list with all possible actions (excluding end of sequence)
         """
         pass
 
-    def get_fixed_policy_output(self):
+    def actions2indices(
+        self, actions: TensorType["batch_size", "action_dim"]
+    ) -> TensorType["batch_size"]:
         """
-        Defines the structure of the output of the policy model, from which an
-        action is to be determined or sampled, by returning a vector with a fixed
-        random policy. As a baseline, the fixed policy is uniform over the
-        dimensionality of the action space.
+        Returns the corresponding indices in the action space of the actions in a batch.
         """
-        return np.ones(len(self.action_space))
+        # Expand the action_space tensor: [batch_size, d_actions_space, action_dim]
+        action_space = torch.unsqueeze(self.action_space_torch, 0).expand(
+            actions.shape[0], -1, -1
+        )
+        # Expand the actions tensor: [batch_size, d_actions_space, action_dim]
+        actions = torch.unsqueeze(actions, 1).expand(-1, self.action_space_dim, -1)
+        # Take the indices at the d_actions_space dimension where all the elements in
+        # the action_dim dimension are True
+        return torch.where(torch.all(actions == action_space, dim=2))[1]
 
-    def get_max_traj_len(
+    def get_mask_invalid_actions_forward(
         self,
-    ):
-        return 1e3
+        state: Optional[List] = None,
+        done: Optional[bool] = None,
```

Comment:
What do you mean by "boilerplate with setting state/done to self values"?

Reply:
I mean something like:
```
if state is None:
    state = self.state.copy()
if done is None:
    done = self.done
```
Which basically should be the case for every env, right? I was thinking if this can't be somehow extracted so that it doesn't have to be repeated in every child class. One option would be to have two methods, `get_mask_invalid_actions_forward` that actually can have optional arguments, and the only thing it does is sets them to `self.{state, done}` and calls `_get_mask_invalid_actions_forward`, which assumes they are not None and has actual logic - but I might be overthinking this and it might be an overkill.

Anyway, it's not a big issue, and I am ok with it being as is.

Reply:
Ah I see! I have actually been thinking of something like this too and I was wondering if there is an elegant way to handle this properly through inheritance. Since this is not crucial right now, I'd leave it for later, but it's a very good point and I agree.